### ISSUE TITLE
Very bad performance on JavaScriptCore

### ISSUE THREAD

**Issue by p0358** [markedjs/marked#2863]
**Marked version:** 5.1.0 (but also happened on 4.x)

**Describe the bug**
Performance of `marked` library is VERY bad on JavaScriptCore (what backs Safari's JS) compared to V8:
- on Ultralight (embedded WebKit/Safari-based browser) it takes 4 seconds to parse 80 KB of simple Markdown
- on bun.js (JSC-based alternative to node.js) it takes 1.5 seconds to parse the same file

On Chromium with your official demo site, it takes like 35-50 ms to parse the same file, which makes up for a very drastic difference. Also the tests are done on Ryzen 7950X to underline the scale of the issue.

Now Safari's profiler isn't the best, but I'm gonna still paste this screenshot:
![image](https://github.com/markedjs/marked/assets/5182588/e13420ca-4057-4635-9ae2-dc2c201d423e)

Also the file (it's just a simple anonymized changelog, there should be nothing fancy hidden in):
[test.md](https://github.com/markedjs/marked/files/11911404/test.md)

**To Reproduce**

Easiest to reproduce it with bun.js (you can use WSL if on Windows):

```console
bun init
bun install marked
time bun run index.js
```

index.js:
```javascript
import * as fs from "fs";
import { marked } from "marked";

const renderer = new marked.Renderer();
renderer.link = (href, title, text) => `<a target="_external" onclick="return window.onLinkClick(this)" href="${ href }" ${title? 'title="'+title+'"' : ''}>${ text }</a>`;
renderer.heading = (text, level) => `${level === 1 ? '</div><div class="changelogEntry">' : ''}<h${level}>${text}</h${level}>`;
const data = fs.readFileSync("./test.md").toString();

marked(data, { renderer }, (error, parseResult) => {
    if (error)
        throw error;
    console.log(parseResult.length);
});
```

```console
$ time bun run index.js
real    0m1.577s
user    0m1.546s
sys     0m0.031s
```

**Expected behavior**

with node.js using V8:
```console
$ time node index.js
real    0m0.071s
user    0m0.059s
sys     0m0.000s
```

V8 is 22x-55x faster than JavaScriptCore with Marked, which is abnormal situation, not normally exhibited by other code or libraries.

**Comment by p0358** [comment]
On an unrelated note, I when running this repro, I noticed that the output length every time the thing is ran is different, which is also very strange (on both bun and node):
![image](https://github.com/markedjs/marked/assets/5182588/559ef4e2-6ff1-4e19-a552-2224fd34fbd0)

**Comment by p0358** [comment]
Also likely related to #509 which was closed, but actually apparently nothing changed since then...

**Comment by p0358** [comment]
Also the repro customizes renderer with custom `link` and `heading` functions, commenting those out does not seem to affect the result at all.

Also here's repro comparison with CommonMark library:

```javascript
import * as fs from "fs";
import * as commonmark from "commonmark";

const data = fs.readFileSync("./test.md").toString();

const reader = new commonmark.Parser();
const writer = new commonmark.HtmlRenderer();
const parsed = reader.parse(data);
const result = writer.render(parsed);
console.log(result.length);
```

```console
$ time bun run commonmark.js
real    0m0.107s
user    0m0.010s
sys     0m0.062s
```

```console
$ time node commonmark.js
real    0m0.246s
user    0m0.039s
sys     0m0.065s
```

As we can see, with commonmark, JavaScriptCore is actually ~2x faster instead!!!

**Comment by UziTech** [comment]
I don't understand why you put this issue on marked. We don't do anything different based on the engine. This seems like an issue that should be reported to safari so they can improve it.

**Comment by p0358** [comment]
I know, but at the same time it's possible that some unoptimized RegExp expressions used here may exhibit such behavior on JSC, which might be possible to work around, perhaps.

But now I see it's been reported by Bun's author here to Safari authors, so I'm pasting the link for reference: https://bugs.webkit.org/show_bug.cgi?id=258706 (to see what'll be their conclusion after they get to it)

**Comment by UziTech** [comment]
We use the issues as a backlog for what needs to get done. If this doesn't include any actionable information it will be closed as stale in a couple weeks.

If you do find any RegExps that could be optimized, a PR would be nice 😁👍

**Comment by Trapfether** [comment]
I spent some time on this today. According to updates on the webkit bug referenced by @p0358, The performance hits are due to several RegEX being interpreted rather than JIT’ed by JavaScriptCore. I attempted to work around possible reasons for this nonoptimal handling.

I tested both hypothesis presented by Michael Saboff at the webkit team, and went so far as to remove all instances of case insensitivity and Unicode handling (just to see if it was even theoretically related) and the performance was unchanged. This does establish that the issue is solely within the Engine itself and is not something that can be worked around or optimized on the part of the library.
